From 4dc4fc114efa30dd089a6a1dac6a6a32d8bb4879 Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Tue, 7 May 2019 19:58:10 +0200
Subject: [PATCH 4/5] [DNM] Squash of lineage-16.0-android-9.0.0_r37

Squashed commit of the following:

commit f8a5ce9945240777b0d5cd2ba68fdda48e1484a8
Merge: 04c6c2121 80e24410b
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Wed Mar 13 18:04:25 2019 +0000

    Merge cherrypicks of [6716922, 6716923, 6716413, 6717023, 6717024, 6716716, 6715859, 6717160, 6717161, 6717162, 6717163, 6716295, 6717141, 6717181, 6717183, 6717184, 6717185, 6714937, 6717028, 6716717, 6716927, 6717200, 6717029, 6717030, 6717031, 6717032, 6717033, 6716928, 6717034, 6717035, 6716929, 6717201, 6716930, 6712377, 6712378, 6716643, 6717164, 6712379] into pi-qpr3-release

    Change-Id: Ida0fb2e062df664494ba6be312ae6131018fe536

commit 80e24410b7c42288f02869543555e682d88213f7
Author: Ajay Panicker <apanicke@google.com>
Date:   Fri Dec 14 14:55:02 2018 -0800

    DO NOT MERGE: Use a weak pointer to deliver updates to AVRCP devices.

    If a device disconnects right before a update message gets queued, the
    device becomes null and there is a crash when the callback for the
    update executes on the disconnected device. This patch switches the
    device reference from being Unretained to using a weak pointer so that
    the callback just doesn't execute if the device is disconnected.

    Bug: 120431125
    Bug: 120445479
    Test: Use the same test as b/120477414 as that bug causes a disconnect
    at the same time as a media update.

    Change-Id: I1dcc08e5c9866106e7ec0dad52505e34b42da600
    (cherry picked from commit f083d1e076ea97e6feaa363f03dab3656bd03ee0)

commit 04c6c2121729853e02a762dba4be1a17d4b325b5
Merge: 0db97596f 2a1ee659e
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Wed Feb 27 04:02:33 2019 +0000

    Snap for 5339334 from 2a1ee659eaa5eb6a03d963f2b8e3ded860a1cc52 to pi-qpr3-release

    Change-Id: I420f5037d8f471edb35b369a0cad6bc5a5728d0c

commit 2a1ee659eaa5eb6a03d963f2b8e3ded860a1cc52
Merge: fdd0bb033 827bd08e1
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Sat Feb 23 07:30:06 2019 +0000

    Merge "Fix mtu assignment with correct value" into pi-dev

commit fdd0bb033f3c43da461e133bb1cc4e6eed5adb38
Merge: 80c748c40 87b6a2075
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Fri Feb 22 05:04:53 2019 +0000

    Merge "Do not close dump file descriptor" into pi-dev

commit 80c748c4073bb428ebaad26c0ea2ca36bc378d79
Merge: 5980eefcf 672c824ef
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Thu Feb 21 23:44:19 2019 +0000

    Merge "Fix for Bluetooth device name is resetting to default name after reboot" into pi-dev

commit 0db97596f4b6ca3016921e8c1c02f4979c394382
Merge: dfd381e10 5980eefcf
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Wed Feb 13 04:03:33 2019 +0000

    Snap for 5304822 from 5980eefcfb684c4f2006be9a767d90fce99973c0 to pi-qpr3-release

    Change-Id: I2c43d4e8ced654fa07d6a197ec7bc9d5a828e561

commit 5980eefcfb684c4f2006be9a767d90fce99973c0
Merge: e14f845fa 850050038
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Feb 8 12:42:30 2019 -0800

    [automerger skipped] Merge "DO NOT MERGE Separate SDP procedure from bonding state (1/2)" into oc-dev am: 99495df519 -s ours am: 8d550cccd6 -s ours
    am: 8500500386 -s ours
    am skip reason: subject contains skip directive

    Change-Id: I4f30022298edd7521c895bdceb743662e915b229

commit e14f845fac6674e8131bdb2b2b72f7c9b1ef7f4a
Merge: 02bf171fb feeb53795
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Feb 8 12:41:07 2019 -0800

    [automerger skipped] DO NOT MERGE Separate SDP procedure from bonding state (1/2) am: 0989441858 -s ours am: a6d6406255 -s ours
    am: feeb53795a -s ours
    am skip reason: subject contains skip directive

    Change-Id: Ie86e28c3ee0c779e9f4df6a1279750e67765eec9

commit 8500500386f463c1acbda3875d406b104ba97ffc
Merge: feeb53795 8d550cccd
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Feb 8 12:38:28 2019 -0800

    [automerger skipped] Merge "DO NOT MERGE Separate SDP procedure from bonding state (1/2)" into oc-dev am: 99495df519 -s ours
    am: 8d550cccd6 -s ours
    am skip reason: change_id I023713e07308bfc0e5bb8d67f386bcc50f6a0f85 with SHA1 f3681c8616 is in history

    Change-Id: Iaf7c1aadcfd65ed6a82b0c3bc98ff121046f70d2

commit feeb53795ad15ab061961c131ad8924ddce30cb3
Merge: 0cd9890f7 a6d640625
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Feb 8 12:37:26 2019 -0800

    [automerger skipped] DO NOT MERGE Separate SDP procedure from bonding state (1/2) am: 0989441858 -s ours
    am: a6d6406255 -s ours
    am skip reason: change_id I023713e07308bfc0e5bb8d67f386bcc50f6a0f85 with SHA1 f3681c8616 is in history

    Change-Id: Ie5679c43d4e59f3028a115558763baae8d31ec73

commit 8d550cccd61e1323801af196a3e6b638b96ee361
Merge: a6d640625 99495df51
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Feb 8 12:34:11 2019 -0800

    [automerger skipped] Merge "DO NOT MERGE Separate SDP procedure from bonding state (1/2)" into oc-dev
    am: 99495df519 -s ours
    am skip reason: change_id I023713e07308bfc0e5bb8d67f386bcc50f6a0f85 with SHA1 0989441858 is in history

    Change-Id: I6fb78cdd8f4c748d6cafa85bdb1f8f60d676373b

commit a6d6406255c45c0889f41f236b9ac71b093d8917
Merge: d73d5ace0 098944185
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Feb 8 12:33:37 2019 -0800

    [automerger skipped] DO NOT MERGE Separate SDP procedure from bonding state (1/2)
    am: 0989441858 -s ours
    am skip reason: change_id I023713e07308bfc0e5bb8d67f386bcc50f6a0f85 with SHA1 edd7e731ed is in history

    Change-Id: Ied8e02f8c14ed6190cc9baebf688359d682b6d00

commit 99495df519892a7d8f484319edbdbc3519514005
Merge: c59317a10 098944185
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Fri Feb 8 20:19:42 2019 +0000

    Merge "DO NOT MERGE Separate SDP procedure from bonding state (1/2)" into oc-dev

commit 02bf171fb24800a1728c7b20a2573aa2039b35d2
Merge: c3ea07116 0cd9890f7
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Feb 8 00:14:17 2019 -0800

    Merge changes from topic "am-43952131-a4db-4e42-bfef-2d44a29b3fac" into oc-dev am: c59317a10a am: d73d5ace07
    am: 0cd9890f74

    Change-Id: I80a5aea488949c646baf1e7f514878ec1cba38e3

commit 0cd9890f74fd77c84b4a6e79f6c15ce2a2369c29
Merge: d59c9ec72 d73d5ace0
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Feb 8 00:10:13 2019 -0800

    Merge changes from topic "am-43952131-a4db-4e42-bfef-2d44a29b3fac" into oc-dev am: c59317a10a
    am: d73d5ace07

    Change-Id: I276abc4b3d069c7700e37cf6c2a34df380b07035

commit d73d5ace07196c0e3547968f6e3fdee23790904b
Merge: c87e949d5 c59317a10
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Feb 8 00:06:25 2019 -0800

    Merge changes from topic "am-43952131-a4db-4e42-bfef-2d44a29b3fac" into oc-dev
    am: c59317a10a

    Change-Id: I424449c094db3c75d1f76f7d397b7b510d1a8959

commit c59317a10a892bb0751c7190e5c8753fcf3b629e
Merge: 4ac889b78 bbe8c6125
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Fri Feb 8 07:54:10 2019 +0000

    Merge changes from topic "am-43952131-a4db-4e42-bfef-2d44a29b3fac" into oc-dev

    * changes:
      [automerger] DO NOT MERGE Separate SDP procedure from bonding state (1/2) am: edd7e731ed am: 279c2a1910 am: c29c3aa408 am: 70ab44a424 skipped: 4e26a1fa5d
      [automerger] DO NOT MERGE Separate SDP procedure from bonding state (1/2) am: edd7e731ed am: 279c2a1910 am: c29c3aa408 am: 70ab44a424
      [automerger] DO NOT MERGE Separate SDP procedure from bonding state (1/2) am: edd7e731ed am: 279c2a1910 am: c29c3aa408
      [automerger] DO NOT MERGE Separate SDP procedure from bonding state (1/2) am: edd7e731ed am: 279c2a1910
      [automerger] DO NOT MERGE Separate SDP procedure from bonding state (1/2) am: edd7e731ed
      DO NOT MERGE Separate SDP procedure from bonding state (1/2)

commit c3ea071167b74143e7b7a9e8014312718e969922
Merge: dc9d0d8f1 d59c9ec72
Author: Hansong Zhang <hsz@google.com>
Date:   Thu Feb 7 16:50:16 2019 -0800

    Merge "btm_proc_smp_cback: Don't access p_dev_rec if freed" into oc-dev am: 4ac889b785 am: c87e949d53
    am: d59c9ec72e

    Change-Id: I31957dc7e323364de3529b3375835dd4cb7a501d

commit d59c9ec72ed18f808c9658e1a34be2ea38b9cf14
Merge: c48874b71 c87e949d5
Author: Hansong Zhang <hsz@google.com>
Date:   Thu Feb 7 16:46:11 2019 -0800

    Merge "btm_proc_smp_cback: Don't access p_dev_rec if freed" into oc-dev am: 4ac889b785
    am: c87e949d53

    Change-Id: I6875480db5a714d721d20ead2111627f4ab5a68e

commit dc9d0d8f17573a862fcec24b0254af880cace290
Merge: 9028a3164 c48874b71
Author: Ugo Yu <ugoyu@google.com>
Date:   Thu Feb 7 16:44:52 2019 -0800

    [automerger skipped] Merge "DO NOT MERGE Separate SDP procedure from bonding state (1/2)" into oc-mr1-dev
    am: c48874b71b -s ours
    am skip reason: subject contains skip directive

    Change-Id: Ibcaf4a7d743ec0cfeebb8aecc7bab2bf3b88615d

commit 9028a316409f098a2c8a614f6e5d7ab1f02b767d
Merge: 7799b0915 f3681c861
Author: Ugo Yu <ugoyu@google.com>
Date:   Thu Feb 7 16:44:14 2019 -0800

    [automerger skipped] DO NOT MERGE Separate SDP procedure from bonding state (1/2)
    am: f3681c8616 -s ours
    am skip reason: subject contains skip directive

    Change-Id: I6a7d8107caa959212e0d93e708dd87851199a2c8

commit c87e949d537bba31ff6b8c7fe4e4eac541b3769a
Merge: 9b0e9a32b 4ac889b78
Author: Hansong Zhang <hsz@google.com>
Date:   Thu Feb 7 16:42:12 2019 -0800

    Merge "btm_proc_smp_cback: Don't access p_dev_rec if freed" into oc-dev
    am: 4ac889b785

    Change-Id: I3a99684487593468b89948aa9d3be99e5ed705f4

commit 4ac889b785fe4cc15ac3bc8039881135547b51b8
Merge: e14580597 953dd2795
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Fri Feb 8 00:31:40 2019 +0000

    Merge "btm_proc_smp_cback: Don't access p_dev_rec if freed" into oc-dev

commit c48874b71bf61c2f77e4027cfcd62d08994c3a77
Merge: 53e961c22 f3681c861
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Fri Feb 8 00:27:55 2019 +0000

    Merge "DO NOT MERGE Separate SDP procedure from bonding state (1/2)" into oc-mr1-dev

commit 7799b0915e16968c6ad34319d7180b0700c36541
Merge: 8c5cac9ac dc14eb8ba
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Thu Feb 7 23:42:23 2019 +0000

    Merge "DO NOT MERGE Separate SDP procedure from bonding state (1/2)" into pi-dev

commit 8c5cac9acfcece105640fe28279d85087806da71
Merge: 812588ad7 53e961c22
Author: Hansong Zhang <hsz@google.com>
Date:   Thu Feb 7 13:25:32 2019 -0800

    Merge changes from topic "am-5380790e-42fb-4784-96c0-4412e4fdccd0" into oc-dev am: e145805974 am: 9b0e9a32bb
    am: 53e961c220

    Change-Id: I9fc3a00e1b38236aaa1016ef0c422fe5303f407b

commit 53e961c220c8e70578114f400ae0770b93a08755
Merge: a99a5f342 9b0e9a32b
Author: Hansong Zhang <hsz@google.com>
Date:   Thu Feb 7 13:21:26 2019 -0800

    Merge changes from topic "am-5380790e-42fb-4784-96c0-4412e4fdccd0" into oc-dev am: e145805974
    am: 9b0e9a32bb

    Change-Id: I02f1344805f748024dc28e05fd0afe67a6afb61a

commit 9b0e9a32bbce9d1cdc30b3b621b5d5cfce4bd37d
Merge: 04f9bde9e e14580597
Author: Hansong Zhang <hsz@google.com>
Date:   Thu Feb 7 13:17:35 2019 -0800

    Merge changes from topic "am-5380790e-42fb-4784-96c0-4412e4fdccd0" into oc-dev
    am: e145805974

    Change-Id: Iaa78778cedd6e04d3cf7d009b81a9599658e6583

commit e145805974baed559dfe65efb5da703aabe50ca9
Merge: 8ea254e22 ade3e18ab
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Thu Feb 7 21:07:56 2019 +0000

    Merge changes from topic "am-5380790e-42fb-4784-96c0-4412e4fdccd0" into oc-dev

    * changes:
      [automerger] DO NOT MERGE btm_proc_smp_cback: Don't access p_dev_rec if freed am: 74c6d501ce am: 059e3c77e2 am: a244a4072c am: 90265d4ee0 skipped: 84ba34d57a
      [automerger] DO NOT MERGE btm_proc_smp_cback: Don't access p_dev_rec if freed am: 74c6d501ce am: 059e3c77e2 am: a244a4072c am: 90265d4ee0
      [automerger] DO NOT MERGE btm_proc_smp_cback: Don't access p_dev_rec if freed am: 74c6d501ce am: 059e3c77e2 am: a244a4072c
      [automerger] DO NOT MERGE btm_proc_smp_cback: Don't access p_dev_rec if freed am: 74c6d501ce am: 059e3c77e2
      [automerger] DO NOT MERGE btm_proc_smp_cback: Don't access p_dev_rec if freed am: 74c6d501ce
      DO NOT MERGE btm_proc_smp_cback: Don't access p_dev_rec if freed

commit dfd381e1083f81324d50a0c11df9a45e5b7f743c
Merge: 70ea66638 812588ad7
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Wed Feb 6 04:13:57 2019 +0000

    Snap for 5285806 from 812588ad71bf672864ea4051a32419f15f2bff69 to pi-qpr3-release

    Change-Id: Iedc2d3cb7604466be6f38f1bac4604f01f778e4f

commit 812588ad71bf672864ea4051a32419f15f2bff69
Merge: 6c0f22f32 a99a5f342
Author: Hansong Zhang <hsz@google.com>
Date:   Sat Feb 2 00:13:14 2019 -0800

    Merge changes from topic "am-cdd47550-8877-443a-826f-db2b25d750ce" into oc-dev am: 8ea254e227 am: 04f9bde9ea
    am: a99a5f342a

    Change-Id: I84c36724790660df845bb3a5ade56f792601ca0d

commit a99a5f342a90273161a82c3966856dfe785d0755
Merge: ec78d7470 04f9bde9e
Author: Hansong Zhang <hsz@google.com>
Date:   Sat Feb 2 00:09:09 2019 -0800

    Merge changes from topic "am-cdd47550-8877-443a-826f-db2b25d750ce" into oc-dev am: 8ea254e227
    am: 04f9bde9ea

    Change-Id: I2f9ab4cc670a7faa7305fe110a964168bd4c40d5

commit 04f9bde9eafe79f652c995f8a5965341eb445ba1
Merge: 82365b0e8 8ea254e22
Author: Hansong Zhang <hsz@google.com>
Date:   Sat Feb 2 00:05:14 2019 -0800

    Merge changes from topic "am-cdd47550-8877-443a-826f-db2b25d750ce" into oc-dev
    am: 8ea254e227

    Change-Id: Ifc5a20abe1c8091f4850cd3d75f9ecfb4474f11e

commit 8ea254e227b78492b5f9968dc91118139b3ba97f
Merge: 356edb433 31dd859b0
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Sat Feb 2 07:52:13 2019 +0000

    Merge changes from topic "am-cdd47550-8877-443a-826f-db2b25d750ce" into oc-dev

    * changes:
      [automerger] DO NOT MERGE process_l2cap_cmd: Fix OOB am: 38f07a3c93 am: 14f6578d9e am: 53e323b2af am: d0584f3dcf skipped: 55b702e6c4
      [automerger] DO NOT MERGE process_l2cap_cmd: Fix OOB am: 38f07a3c93 am: 14f6578d9e am: 53e323b2af am: d0584f3dcf
      [automerger] DO NOT MERGE process_l2cap_cmd: Fix OOB am: 38f07a3c93 am: 14f6578d9e am: 53e323b2af
      [automerger] DO NOT MERGE process_l2cap_cmd: Fix OOB am: 38f07a3c93 am: 14f6578d9e
      [automerger] DO NOT MERGE process_l2cap_cmd: Fix OOB am: 38f07a3c93
      DO NOT MERGE process_l2cap_cmd: Fix OOB

commit 6c0f22f324ed0bdf9dea3e803e5ee6176d03fdb4
Merge: f4aa47002 ec78d7470
Author: Hansong Zhang <hsz@google.com>
Date:   Fri Feb 1 17:45:30 2019 -0800

    resolve merge conflicts of ec78d74706c3e81f91eee53e3d9f959f66e5d77f to pi-dev

    Bug: None
    Test: I solemnly swear I tested this conflict resolution.
    Change-Id: Id658b3485fdc0025bc44850be9f23bb2d2146d9b

commit ec78d74706c3e81f91eee53e3d9f959f66e5d77f
Merge: a3a57067d 82365b0e8
Author: Hansong Zhang <hsz@google.com>
Date:   Fri Feb 1 15:39:08 2019 -0800

    Merge "process_l2cap_cmd: Fix OOB" into oc-dev am: 356edb4333
    am: 82365b0e8d

    Change-Id: I4f04fad60e84785f474390e473d2fcf19f66a044

commit 82365b0e8da13e1b5146954a9d71b5a72d7f3980
Merge: 154230b83 356edb433
Author: Hansong Zhang <hsz@google.com>
Date:   Fri Feb 1 15:33:09 2019 -0800

    Merge "process_l2cap_cmd: Fix OOB" into oc-dev
    am: 356edb4333

    Change-Id: I0e1800587513bdb39e0b6eff7e46254470ab2def

commit 356edb43334e38130ba5dd633725e8c5a1de6fe3
Merge: 19460901d 94fd011bc
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Fri Feb 1 23:18:22 2019 +0000

    Merge "process_l2cap_cmd: Fix OOB" into oc-dev

commit 70ea666387f3bbdfe37604d4f8f4947f71a517e1
Merge: 4a0e17fd7 f4aa47002
Author: android-build-team Robot <android-build-team-robot@google.com>
Date:   Wed Jan 30 04:12:53 2019 +0000

    Snap for 5268384 from f4aa4700287798485eed1f356a32422d87f3569e to pi-qpr3-release

    Change-Id: Iea4693ebfb459e1a9e25fc8a49227d8fabe57e63

commit dc14eb8ba01ee20dc44bcba7b3c0a17550e756fa
Author: Ugo Yu <ugoyu@google.com>
Date:   Tue Oct 16 14:53:35 2018 +0800

    DO NOT MERGE Separate SDP procedure from bonding state (1/2)

    - Do not stay in bonding state if the device is paried but still
      discovering service.
    - Report BOND_BONDED to Java after authentication is completed.
    - Report empty UUID to Java if a classic Bluetooth device SDP
      failed while pairing.
    - Hold BOND_BONDED intent util SDP is findished.
    - Only accept profile connection for the device is at bonded
      state. Any attempt to connect while bonding would potentially
      lead to an unauthorized connection.

    Bug: 79703832
    Test: runtest bluetooth, regression test
    Change-Id: I023713e07308bfc0e5bb8d67f386bcc50f6a0f85
    (cherry picked from commit 122e115b87fe98ca5e5e65b9765c146f9e52b65e)

commit f4aa4700287798485eed1f356a32422d87f3569e
Merge: 4a0e17fd7 a3a57067d
Author: Hansong Zhang <hsz@google.com>
Date:   Tue Jan 29 17:02:43 2019 -0800

    Merge "btm_ble_multi_adv: Check data length in HCI interface" into oc-dev am: 19460901d8 am: 154230b832
    am: a3a57067d0

    Change-Id: Ib76648858de768c8b3db411eda84efe7abbd420f

commit a3a57067d058a50d5e77bf67d9e10d5f75ad03da
Merge: 56e54c0f0 154230b83
Author: Hansong Zhang <hsz@google.com>
Date:   Tue Jan 29 16:56:34 2019 -0800

    Merge "btm_ble_multi_adv: Check data length in HCI interface" into oc-dev am: 19460901d8
    am: 154230b832

    Change-Id: I69aba9f5350a2b4510e49494839bbff6369c8b4a

commit 154230b83208210d07266523fd0fa03311818628
Merge: af8cf9877 19460901d
Author: Hansong Zhang <hsz@google.com>
Date:   Tue Jan 29 16:49:16 2019 -0800

    Merge "btm_ble_multi_adv: Check data length in HCI interface" into oc-dev
    am: 19460901d8

    Change-Id: If4f3b40817ff57bdae4777ae330854a9119ae0b7

commit 19460901d897eeefccc99548ecfaa063d3335046
Merge: 3fd73e4ad a99fe8a17
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Wed Jan 30 00:25:28 2019 +0000

    Merge "btm_ble_multi_adv: Check data length in HCI interface" into oc-dev

commit ade3e18abc6c2592fec0f3845468fd005c8ad227
Merge: 3fd73e4ad 84ba34d57
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Jan 22 21:47:38 2019 +0000

    [automerger] DO NOT MERGE btm_proc_smp_cback: Don't access p_dev_rec if freed am: 74c6d501ce am: 059e3c77e2 am: a244a4072c am: 90265d4ee0 skipped: 84ba34d57a

    Change-Id: I73f54778128ee9bf1ed46c55bbd545b29ed2dc54

commit 84ba34d57a25b2e78c811bdfabfa5daad1948e35
Merge: dff13d810 90265d4ee
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Jan 22 21:47:36 2019 +0000

    [automerger] DO NOT MERGE btm_proc_smp_cback: Don't access p_dev_rec if freed am: 74c6d501ce am: 059e3c77e2 am: a244a4072c am: 90265d4ee0

    Change-Id: I080739b77c52af5ff54bfc4e8a20cf8fd52b235b

commit 90265d4ee0fbaae6bf730f975970058ab4e963b2
Merge: 2ebe3d52b a244a4072
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Jan 22 21:47:35 2019 +0000

    [automerger] DO NOT MERGE btm_proc_smp_cback: Don't access p_dev_rec if freed am: 74c6d501ce am: 059e3c77e2 am: a244a4072c

    Change-Id: Ic43337c91c1cdcb9eaea22311cd7205dc05dcfa2

commit 953dd279502980b1d8d30656eb78c6445a6e31f7
Author: Hansong Zhang <hsz@google.com>
Date:   Wed Jan 9 18:18:17 2019 -0800

    btm_proc_smp_cback: Don't access p_dev_rec if freed

    In btm_proc_smp_cback(), return after p_dev_rec is freed in the middle
    to prevent use after free

    Bug: 120612744
    Test: Use ASAN build; connect to a LE device and wait for timeout
    Change-Id: Ic9d0eaeb62a1a1b24884146ca82f4104fabc5bac

commit a244a4072c4c87f35bd0ae2ecef76a809beb7293
Merge: 097ecf3d8 059e3c77e
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Jan 22 21:47:33 2019 +0000

    [automerger] DO NOT MERGE btm_proc_smp_cback: Don't access p_dev_rec if freed am: 74c6d501ce am: 059e3c77e2

    Change-Id: I96de72b97a23eebad116c98899f59f399614cff7

commit 059e3c77e2fb12be40c192f3885ad018ffb4ad7f
Merge: 85b4574a3 74c6d501c
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Jan 22 21:47:32 2019 +0000

    [automerger] DO NOT MERGE btm_proc_smp_cback: Don't access p_dev_rec if freed am: 74c6d501ce

    Change-Id: Iad8449f422afb55305d3f1f2a148a4122c49c7d8

commit 74c6d501ce55e7bbce4129fae26bd0b5f802a7fc
Author: Hansong Zhang <hsz@google.com>
Date:   Tue Jan 22 13:46:47 2019 -0800

    DO NOT MERGE btm_proc_smp_cback: Don't access p_dev_rec if freed

    In btm_proc_smp_cback(), return after p_dev_rec is freed in the middle
    to prevent use after free

    Bug: 120612744
    Test: Use ASAN build; connect to a LE device and wait for timeout
    Change-Id: I09aa1cf1d1c835146b62d0f4989aeedfb885d95b

commit 31dd859b0133d6c4dadec2ddf274548916f06250
Merge: 3fd73e4ad 55b702e6c
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Jan 22 18:47:24 2019 +0000

    [automerger] DO NOT MERGE process_l2cap_cmd: Fix OOB am: 38f07a3c93 am: 14f6578d9e am: 53e323b2af am: d0584f3dcf skipped: 55b702e6c4

    Change-Id: If1fb97bc56d2ed652f56f1f962aea1d00843543e

commit 55b702e6c4ba1e0f4c9a3c1aaa30cfcb2e832b20
Merge: dff13d810 d0584f3dc
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Jan 22 18:47:22 2019 +0000

    [automerger] DO NOT MERGE process_l2cap_cmd: Fix OOB am: 38f07a3c93 am: 14f6578d9e am: 53e323b2af am: d0584f3dcf

    Change-Id: I101a465864f054989085bba0ccf2fc633445f356

commit d0584f3dcfd4d2be40513dd0e33ce226fc7fb69f
Merge: 2ebe3d52b 53e323b2a
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Jan 22 18:47:20 2019 +0000

    [automerger] DO NOT MERGE process_l2cap_cmd: Fix OOB am: 38f07a3c93 am: 14f6578d9e am: 53e323b2af

    Change-Id: I9a919a3168f0d37834a14778c3f24f1e5f417685

commit 53e323b2afd22c1410020d8e3d402381555c7436
Merge: 097ecf3d8 14f6578d9
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Jan 22 18:47:18 2019 +0000

    [automerger] DO NOT MERGE process_l2cap_cmd: Fix OOB am: 38f07a3c93 am: 14f6578d9e

    Change-Id: I1df2130c25d9399d2c6ebc47bc0b8ec127994b89

commit 14f6578d9e81ce3668d0da8e2349875847014d2e
Merge: 85b4574a3 38f07a3c9
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Jan 22 18:47:16 2019 +0000

    [automerger] DO NOT MERGE process_l2cap_cmd: Fix OOB am: 38f07a3c93

    Change-Id: I89bb716ce51a1d98147c0df527174b4934999347

commit 38f07a3c93143ca31229f0caa5b1a270dc1f5401
Author: Hansong Zhang <hsz@google.com>
Date:   Fri Jan 18 11:51:00 2019 -0800

    DO NOT MERGE process_l2cap_cmd: Fix OOB

    Bug: 119870451
    Test: POC
    Change-Id: Ieef322a3ad4cebcaf40e5388584d3a04a4761d2e

commit 94fd011bc9a72081cc691ed7d6e6eec42e9f4539
Author: Hansong Zhang <hsz@google.com>
Date:   Mon Jan 14 14:59:35 2019 -0800

    process_l2cap_cmd: Fix OOB

    Bug: 119870451
    Test: POC
    Change-Id: I2f5e7fedd9aed96c4ffc55af79fdac61c2e5b087
    Merged-In: I5131bbf9cda6248fdbbc4bb91916b2fe3731246e

commit a99fe8a175a6d209e741871544ae3f857c8a7cbb
Author: Hansong Zhang <hsz@google.com>
Date:   Wed Jan 16 12:33:26 2019 -0800

    btm_ble_multi_adv: Check data length in HCI interface

    For BleAdvertiserVscHciInterfaceImpl and
    BleAdvertiserLegacyHciInterfaceImpl, the maximum size of scan response
    and advertising packet data length should be BTM_BLE_AD_DATA_LEN (31).

    Bug: 121145627
    Test: POC
    Change-Id: I7653a6c186b7313ef2b1547bca120b9d41c90140

commit 87b6a207536513a5d7a8886827e2c00763173c81
Author: vich <vich@google.com>
Date:   Mon Dec 10 13:14:03 2018 +0800

    Do not close dump file descriptor

    The dump caller should close the file descriptor(fd), close fd after
    bluetooth dump will cause fd be closed twice.

    Bug:116831525
    Bug:121049917
    Test: Run BT off/on with headset connected test 10000 times.
    Change-Id: I59b60043a0a346de9ec7ebadb91c55d0dcea6e2a
    Merged-In: Id351b5429cb17b48e37313850a363d739d78c4a1

commit 4a0e17fd7ed8c0a7f774324cbb56d6bb9ae2d576
Merge: d36a3367b 56e54c0f0
Author: Stanley Tng <stng@google.com>
Date:   Mon Jan 7 14:50:38 2019 -0800

    [automerger skipped] Merge "DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu" into oc-dev am: 3fd73e4ad2 -s ours am: af8cf98776 -s ours
    am: 56e54c0f07 -s ours
    am skip reason: subject contains skip directive

    Change-Id: Id562d02c179dd9299f06d8d09fef15afc94cfb67

commit d36a3367b9326ef5241f8e24b537da60090c6baa
Merge: c117a1c95 4dc3df74c
Author: Stanley Tng <stng@google.com>
Date:   Mon Jan 7 14:50:24 2019 -0800

    [automerger skipped] [automerger] DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu am: c1fcbd5508 am: 85b4574a31 am: 097ecf3d88 am: 2ebe3d52b0 skipped: dff13d810c am: 47dcb6a458 am: a2e761ac2b
    am: 4dc3df74c5 -s ours
    am skip reason: subject contains skip directive

    Change-Id: Iaac7919f35e29539fb245fe26428b9c378861e83

commit 56e54c0f07f0169ed4a00e0bedd0f69c8a12567e
Merge: 4dc3df74c af8cf9877
Author: Stanley Tng <stng@google.com>
Date:   Mon Jan 7 14:45:06 2019 -0800

    [automerger skipped] Merge "DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu" into oc-dev am: 3fd73e4ad2 -s ours
    am: af8cf98776 -s ours
    am skip reason: change_id I30b7a8af11d3a5f974cb39e06b0e3463bebc8e9a with SHA1 c1fcbd5508 is in history

    Change-Id: Ib82a5c8869f17e93969906a55323e43a79628f4a

commit 4dc3df74c593f9605bd4aa92fde0a60252eab946
Merge: 845fc7c23 a2e761ac2
Author: Stanley Tng <stng@google.com>
Date:   Mon Jan 7 14:44:41 2019 -0800

    [automerger] DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu am: c1fcbd5508 am: 85b4574a31 am: 097ecf3d88 am: 2ebe3d52b0 skipped: dff13d810c am: 47dcb6a458
    am: a2e761ac2b

    Change-Id: Ia655d085c410e45e44efc1fa1bb2737a7cccd310

commit af8cf98776f2136c6dae90bcc1d30b6f0cef8837
Merge: a2e761ac2 3fd73e4ad
Author: Stanley Tng <stng@google.com>
Date:   Mon Jan 7 14:40:43 2019 -0800

    [automerger skipped] Merge "DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu" into oc-dev
    am: 3fd73e4ad2 -s ours
    am skip reason: change_id I30b7a8af11d3a5f974cb39e06b0e3463bebc8e9a with SHA1 c1fcbd5508 is in history

    Change-Id: Ie17711d48b90d8921cf26dc501cdfb776c7ef47e

commit a2e761ac2b348c4d3ff5722b6da08276ad8df497
Merge: 45a3f8a6f 47dcb6a45
Author: Stanley Tng <stng@google.com>
Date:   Mon Jan 7 14:40:22 2019 -0800

    [automerger] DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu am: c1fcbd5508 am: 85b4574a31 am: 097ecf3d88 am: 2ebe3d52b0 skipped: dff13d810c
    am: 47dcb6a458

    Change-Id: Ief422a38b0c559b912038c038edc48854357b3bf

commit 3fd73e4ad26a1b79589d197a8e5ec288110d2f2b
Merge: 47dcb6a45 8f52ed93b
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Mon Jan 7 22:33:36 2019 +0000

    Merge "DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu" into oc-dev

commit c117a1c951c65033987ed51d53407b359204c187
Author: Stanley Tng <stng@google.com>
Date:   Tue Dec 11 14:45:13 2018 -0800

    DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu

    Add check to make sure that data buffer is big enough to read the 2
    bytes for length.

    Also, fix a regression from the previous CL that checks the buffer length
    before doing a memcpy. The previous check is too strict causing valid
    sized buffers to be rejected. The length check is incorrect and off by the header size.

    Bug: 120665616
    Test: Run the SL4A Test for LE CoC, BleCoCTest
    Merged-In: I30b7a8af11d3a5f974cb39e06b0e3463bebc8e9a
    Change-Id: I30b7a8af11d3a5f974cb39e06b0e3463bebc8e9a
    (cherry picked from commit fcb1994de1f6ee34b8dc6804a2b32e20bf138073)
    (cherry picked from commit 1f1d8b97d80d25023c4c7b04d2aa18d367f4158d)

commit 8f52ed93ba0fe67c310473b539d37c7201c83454
Author: Stanley Tng <stng@google.com>
Date:   Tue Dec 11 14:45:13 2018 -0800

    DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu

    Add check to make sure that data buffer is big enough to read the 2
    bytes for length.

    Also, fix a regression from the previous CL that checks the buffer length
    before doing a memcpy. The previous check is too strict causing valid
    sized buffers to be rejected. The length check is incorrect and off by the header size.

    Bug: 120665616
    Test: Run the SL4A Test for LE CoC, BleCoCTest
    Merged-In: I30b7a8af11d3a5f974cb39e06b0e3463bebc8e9a
    Change-Id: I30b7a8af11d3a5f974cb39e06b0e3463bebc8e9a
    (cherry picked from commit fcb1994de1f6ee34b8dc6804a2b32e20bf138073)
    (cherry picked from commit 1f1d8b97d80d25023c4c7b04d2aa18d367f4158d)
    (cherry picked from commit 6b2739f309f7719086eb8201b3e1a35ba60035f4)

commit 47dcb6a45807a74e4a582a1c14a71b63b5e6d1ce
Merge: f44cbb20e dff13d810
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Sun Jan 6 21:36:29 2019 +0000

    [automerger] DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu am: c1fcbd5508 am: 85b4574a31 am: 097ecf3d88 am: 2ebe3d52b0 skipped: dff13d810c

    Change-Id: I92b4d78f5b6a53c863e7ec6d91b4cc32982258f8

commit dff13d810cb144fda864353427faa8b01b4872d1
Merge: 89e9bbb83 2ebe3d52b
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Sun Jan 6 21:36:28 2019 +0000

    [automerger] DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu am: c1fcbd5508 am: 85b4574a31 am: 097ecf3d88 am: 2ebe3d52b0

    Change-Id: I0cbec621cadfaaf9142d427b52a17cd9db3cd08a

commit 2ebe3d52b04dc3be064ffdeafbce18106c8be583
Merge: c75667da9 097ecf3d8
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Sun Jan 6 21:36:27 2019 +0000

    [automerger] DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu am: c1fcbd5508 am: 85b4574a31 am: 097ecf3d88

    Change-Id: I9fd0733ff10442ca2050e440b954a9cb2f574c1a

commit 097ecf3d8806be0ace85b178dbab208d06d6d30d
Merge: 98ced409a 85b4574a3
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Sun Jan 6 21:36:26 2019 +0000

    [automerger] DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu am: c1fcbd5508 am: 85b4574a31

    Change-Id: I40ce009c5868fde902bc29a0af1b62c89f02f158

commit 85b4574a31a98a2d31219e31d62248fe981266d6
Merge: 12d8535d0 c1fcbd550
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Sun Jan 6 21:36:24 2019 +0000

    [automerger] DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu am: c1fcbd5508

    Change-Id: I5812786ed1ac013a273e300c1ddbe3fd26857543

commit c1fcbd5508a75ae3eaf5f311d706d026fee2fe48
Author: Stanley Tng <stng@google.com>
Date:   Tue Dec 11 14:45:13 2018 -0800

    DO NOT MERGE A security fix to check buffer length in l2c_lcc_proc_pdu

    Add check to make sure that data buffer is big enough to read the 2
    bytes for length.

    Also, fix a regression from the previous CL that checks the buffer length
    before doing a memcpy. The previous check is too strict causing valid
    sized buffers to be rejected. The length check is incorrect and off by the header size.

    Bug: 120665616
    Test: Run the SL4A Test for LE CoC, BleCoCTest
    Merged-In: I30b7a8af11d3a5f974cb39e06b0e3463bebc8e9a
    Change-Id: I30b7a8af11d3a5f974cb39e06b0e3463bebc8e9a
    (cherry picked from commit fcb1994de1f6ee34b8dc6804a2b32e20bf138073)
    (cherry picked from commit 1f1d8b97d80d25023c4c7b04d2aa18d367f4158d)
    (cherry picked from commit 6b2739f309f7719086eb8201b3e1a35ba60035f4)

commit 672c824efe6e45ab3ce973ce879a71bc5d795d27
Author: Aiswarya Cyriac <aiswarya.cyriac@intel.com>
Date:   Mon Jul 23 15:17:48 2018 +0530

    Fix for Bluetooth device name is resetting to default name after reboot

    Reason: When the BT device name is updated is getting saved to the
    config data pointer but not to the persistent data (i.e. bt_config.conf).
    So, when the reboot is happening it is not able to get the updated device
    name from the persistent data (i.e. bt_config.conf) as during reboot
    bt_config_flush is not called.

    Fix: Saving the BT Device name to persistent data using btif_config_flush
    once it is set.

    Test: Manual Reboot test
    Bug: 110301897
    Bug: 120631885
    Change-Id: Ie329b475eaaf208ed667f27e271a00e230e4b95c
    Signed-off-by: Aiswarya Cyriac <aiswarya.cyriac@intel.com>
    Signed-off-by: Gaganpreet kaur <gaganpreetx.kaur@intel.com>

commit 827bd08e1e72131fad876b94311c44262082769e
Author: Ted Wang <tedwang@google.com>
Date:   Tue Dec 18 10:31:06 2018 +0800

    Fix mtu assignment with correct value

    Assigned mtu value with correspond avct channel configuration

    Bug: 120524683
    Test: manual
    Change-Id: Idb70c678cef1d704e232aef8706082a48b073ef8

commit f7d26fe4779d580844e6d600f75eead8e50ddd3a
Merge: 6b408d20b 845fc7c23
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Nov 30 13:27:11 2018 -0800

    Add OOB check in avrc_pars_browse_rsp am: f44cbb20e7 am: 45a3f8a6fd
    am: 845fc7c234

    Change-Id: I4e6e5da28d3aaa25c9f5522385ae098b4c8362cc

commit 845fc7c234ca09ef8ed78144b89add4b088c4513
Merge: 8e403606f 45a3f8a6f
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Nov 30 13:23:05 2018 -0800

    Add OOB check in avrc_pars_browse_rsp am: f44cbb20e7
    am: 45a3f8a6fd

    Change-Id: Ibb3ff9d8915ab9129e890b04848dd78891262e28

commit 45a3f8a6fd290bcdb9c4e3d5299d1aee64c1124e
Merge: 9eac53f26 f44cbb20e
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Nov 30 13:13:52 2018 -0800

    Add OOB check in avrc_pars_browse_rsp
    am: f44cbb20e7

    Change-Id: Ieabff141ded21319a946fa0829bdb60cdedd8e9b

commit f44cbb20e7658116472981bac0ffb0305f4a2c04
Author: Ugo Yu <ugoyu@google.com>
Date:   Tue Nov 13 20:03:28 2018 +0800

    Add OOB check in avrc_pars_browse_rsp

    Bug: 111451066
    Test: Manully
    Change-Id: I068d218b8957bb8f053148d252a9119a8def28cc

commit 6b408d20b7c7fc56c041027a5994af58836d4831
Merge: 469606d73 8e403606f
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Thu Nov 29 05:41:21 2018 -0800

    [automerger skipped] [automerger] Fix buffer overflow in btif_dm_data_copy am: d117975904 am: 12d8535d0f am: 98ced409a5 am: c75667da96 skipped: 89e9bbb83c am: 2f53d6ce3c  -s ours am: 9eac53f263  -s ours
    am: 8e403606f1  -s ours

    Change-Id: Ie3788a8388c6a2fb00969e3b56c0637be82dc092

commit 8e403606f1381af0b9623086c6e02537c8a7b448
Merge: 0dce8b528 9eac53f26
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Thu Nov 29 05:35:18 2018 -0800

    [automerger skipped] [automerger] Fix buffer overflow in btif_dm_data_copy am: d117975904 am: 12d8535d0f am: 98ced409a5 am: c75667da96 skipped: 89e9bbb83c am: 2f53d6ce3c  -s ours
    am: 9eac53f263  -s ours

    Change-Id: I2eed02578d152324e7aa9281e55c3066a4645b33

commit 9eac53f2632fb03367542b3ca1be25a6395b7957
Merge: 1087c5291 2f53d6ce3
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Thu Nov 29 05:31:11 2018 -0800

    [automerger skipped] [automerger] Fix buffer overflow in btif_dm_data_copy am: d117975904 am: 12d8535d0f am: 98ced409a5 am: c75667da96 skipped: 89e9bbb83c
    am: 2f53d6ce3c  -s ours

    Change-Id: If82ef8a6331a62d932eeb3dffaee577ee7d7ea25

commit 2f53d6ce3c3bc44e7234007c0c4b205b3aa29c07
Merge: 969b2df3a 89e9bbb83
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu Nov 29 11:52:36 2018 +0000

    [automerger] Fix buffer overflow in btif_dm_data_copy am: d117975904 am: 12d8535d0f am: 98ced409a5 am: c75667da96 skipped: 89e9bbb83c

    Change-Id: Ia431ddd5ad1d6ee86bd6edd1057372b8dbf51d3b

commit 89e9bbb83cd39ca7a1e769259cd81a3b4066b894
Merge: 247070640 c75667da9
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu Nov 29 11:51:42 2018 +0000

    [automerger] Fix buffer overflow in btif_dm_data_copy am: d117975904 am: 12d8535d0f am: 98ced409a5 am: c75667da96

    Change-Id: I0e5f1348f27f0d9981f99cc0897f9dcc9f443bf3

commit c75667da9654d0b7d2f4b35a0a1f0f11b21fae48
Merge: 12557bb99 98ced409a
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu Nov 29 11:51:39 2018 +0000

    [automerger] Fix buffer overflow in btif_dm_data_copy am: d117975904 am: 12d8535d0f am: 98ced409a5

    Change-Id: I258a6e883061d68b24b30e17e03f72d2000e5f3f

commit 98ced409a5cf6e4a98d23ea6ff610669f0584261
Merge: 3f5af0aa6 12d8535d0
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu Nov 29 11:51:37 2018 +0000

    [automerger] Fix buffer overflow in btif_dm_data_copy am: d117975904 am: 12d8535d0f

    Change-Id: I22ea297e564616790fd7e916747cdcea25d2b068

commit 12d8535d0f717d93b40a09cc213c38ecacdea91a
Merge: a236f1607 d11797590
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Thu Nov 29 11:51:34 2018 +0000

    [automerger] Fix buffer overflow in btif_dm_data_copy am: d117975904

    Change-Id: Icbd5b31039dbf3016575f9d6d69b216d76564c96

commit d1179759041eb66baf1b5cd398d69ce58849d848
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Tue Nov 27 17:59:57 2018 +0100

    Fix buffer overflow in btif_dm_data_copy

    When we use a union, we should always define variables as the union type,
    not as one of the field subtypes. If the latter is cast to the union type,
    buffer overflow can happen.

    Bug: 110166268
    Test: compilation
    Change-Id: I473c03b099ad5a326e7a3739f65efd33cf4775bd
    Merged-In: I473c03b099ad5a326e7a3739f65efd33cf4775bd

commit 469606d73e86c3ac3061817657a5de385323c6f8
Merge: 941cd9360 0dce8b528
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Thu Nov 29 01:46:06 2018 -0800

    [automerger skipped] Fix buffer overflow in btif_dm_data_copy am: 969b2df3a0 am: 1087c5291a  -s ours
    am: 0dce8b5287  -s ours

    Change-Id: Ie4a34e3043f232105344fbe6060d175d502f92ee

commit 0dce8b52877abc677efd33c9ec2ea1a1a4b310fc
Merge: 64c6f33e7 1087c5291
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Thu Nov 29 01:39:01 2018 -0800

    [automerger skipped] Fix buffer overflow in btif_dm_data_copy am: 969b2df3a0
    am: 1087c5291a  -s ours

    Change-Id: I40657cf9ce6ae8ffe3d2d568aaeb34668da84292

commit 1087c5291a084498dca06c2f906a59468726f953
Merge: 1062b9422 969b2df3a
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Thu Nov 29 01:33:56 2018 -0800

    Fix buffer overflow in btif_dm_data_copy
    am: 969b2df3a0

    Change-Id: Icc0b739672cf0683edf9bfc5d8244b1ceb87a1b6

commit 969b2df3a08f9a6429f31725a6b057ce53531fe5
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Tue Nov 27 18:22:22 2018 +0100

    Fix buffer overflow in btif_dm_data_copy

    When we use a union, we should always define variables as the union type,
    not as one of the field subtypes. If the latter is cast to the union type,
    buffer overflow can happen.

    Bug: 110166268
    Test: compilation
    Change-Id: I473c03b099ad5a326e7a3739f65efd33cf4775bd
    Merged-In: I473c03b099ad5a326e7a3739f65efd33cf4775bd

commit 941cd936041d990ad76ba61bcd94e65815ec1f23
Merge: ea90417d9 64c6f33e7
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Wed Nov 28 16:47:17 2018 -0800

    [automerger skipped] Fix buffer overflow in btif_dm_data_copy
    am: 64c6f33e7e  -s ours

    Change-Id: I9e2836c4af3225954fc27961431fcc79f96db072

commit 64c6f33e7e3245f0bc2109001893704763a2ff79
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Tue Nov 27 18:22:22 2018 +0100

    Fix buffer overflow in btif_dm_data_copy

    When we use a union, we should always define variables as the union type,
    not as one of the field subtypes. If the latter is cast to the union type,
    buffer overflow can happen.

    Bug: 110166268
    Test: compilation
    Change-Id: I473c03b099ad5a326e7a3739f65efd33cf4775bd
    Merged-In: I473c03b099ad5a326e7a3739f65efd33cf4775bd

commit ea90417d9965aec1c475418ca8f8f305af12de2d
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Tue Nov 27 18:22:22 2018 +0100

    Fix buffer overflow in btif_dm_data_copy

    When we use a union, we should always define variables as the union type,
    not as one of the field subtypes. If the latter is cast to the union type,
    buffer overflow can happen.

    Bug: 110166268
    Test: compilation
    Change-Id: I473c03b099ad5a326e7a3739f65efd33cf4775bd
    Merged-In: I473c03b099ad5a326e7a3739f65efd33cf4775bd

commit 4f54a847d5c30efa7454e79091e9652790baf1dd
Merge: 0b614505e 54f6c84df
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Tue Nov 27 13:29:49 2018 -0800

    [automerger skipped] [automerger] Fix potential usage of freed memory in btif_hl_proc_sdp_query_cfm am: 78508d2c2c am: a236f16071 am: 3f5af0aa65 am: 12557bb999 skipped: 2470706409 am: d2dd0bacbc  -s ours am: 1062b94226  -s ours
    am: 54f6c84dfd  -s ours

    Change-Id: I4cb634d3c98dabfdc6f1d95f6c1ff348f954ddd4

commit 54f6c84dfd170985c6faad235c94bac6f6f690f3
Merge: 6135f0386 1062b9422
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Tue Nov 27 13:24:43 2018 -0800

    [automerger skipped] [automerger] Fix potential usage of freed memory in btif_hl_proc_sdp_query_cfm am: 78508d2c2c am: a236f16071 am: 3f5af0aa65 am: 12557bb999 skipped: 2470706409 am: d2dd0bacbc  -s ours
    am: 1062b94226  -s ours

    Change-Id: Ic8bab7e79d802eca8efda8a613f83e465e2a4e81

commit 1062b94226ac8c64fd0fb2c2a9d22ff99ea55aa7
Merge: 934213ef0 d2dd0bacb
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Tue Nov 27 13:14:37 2018 -0800

    [automerger skipped] [automerger] Fix potential usage of freed memory in btif_hl_proc_sdp_query_cfm am: 78508d2c2c am: a236f16071 am: 3f5af0aa65 am: 12557bb999 skipped: 2470706409
    am: d2dd0bacbc  -s ours

    Change-Id: If276bb160498a352e02bdb6231ac8fed8142aec4

commit d2dd0bacbc1107d98e456f8992ee0629d594c0ba
Merge: 889efd5b9 247070640
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Nov 27 20:09:16 2018 +0000

    [automerger] Fix potential usage of freed memory in btif_hl_proc_sdp_query_cfm am: 78508d2c2c am: a236f16071 am: 3f5af0aa65 am: 12557bb999 skipped: 2470706409

    Change-Id: Id4bfbba911ecb95c728e1daba294fefc9d1de4ce

commit 24707064094d79b41967e09fa2be7777a0321bb0
Merge: 9c2fb57ce 12557bb99
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Nov 27 16:47:48 2018 +0000

    [automerger] Fix potential usage of freed memory in btif_hl_proc_sdp_query_cfm am: 78508d2c2c am: a236f16071 am: 3f5af0aa65 am: 12557bb999

    Change-Id: I1ecbacc502b14733b0f4bd11b057763506b1fd95

commit 12557bb999b231edcd91f0539533982c687b33a0
Merge: c96313fb2 3f5af0aa6
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Nov 27 16:47:45 2018 +0000

    [automerger] Fix potential usage of freed memory in btif_hl_proc_sdp_query_cfm am: 78508d2c2c am: a236f16071 am: 3f5af0aa65

    Change-Id: I98ae5ab9e24acd447c0c72835067db0bc7430371

commit 3f5af0aa6531c90384b7ebc92d3871836cc736c3
Merge: f9606e1d8 a236f1607
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Nov 27 16:47:42 2018 +0000

    [automerger] Fix potential usage of freed memory in btif_hl_proc_sdp_query_cfm am: 78508d2c2c am: a236f16071

    Change-Id: I8615cedf8b9192c46506c54934229089021fe101

commit a236f160718376a688151d073793c67a3ba08f2b
Merge: 9805ed7a7 78508d2c2
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Nov 27 16:47:40 2018 +0000

    [automerger] Fix potential usage of freed memory in btif_hl_proc_sdp_query_cfm am: 78508d2c2c

    Change-Id: If8da202c56ee7deeb7aba67f59b19ef28466f6ae

commit 78508d2c2cf93b4dd5c7aa630b90f5c6283fe53c
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Tue Nov 20 22:31:31 2018 +0100

    Fix potential usage of freed memory in btif_hl_proc_sdp_query_cfm

    Bug: 116222069
    Test: compilation
    Change-Id: Iebe2c500dfc2806ca321fdcd170e20c680619d4d
    Merged-In: Iebe2c500dfc2806ca321fdcd170e20c680619d4d

commit 0b614505eaad0238461d8f99df93f9f0a270821c
Merge: 2cb7fadb9 6135f0386
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Tue Nov 27 08:31:14 2018 -0800

    Fix potential usage of freed memory in btif_hl_proc_sdp_query_cfm am: 889efd5b91 am: 934213ef03
    am: 6135f03864

    Change-Id: I72d0554197d85b6d63b37c09e67083a21b98ef32

commit 6135f03864bf119ee4634f63010639fff3c742b8
Merge: 40cbdde49 934213ef0
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Tue Nov 27 08:23:09 2018 -0800

    Fix potential usage of freed memory in btif_hl_proc_sdp_query_cfm am: 889efd5b91
    am: 934213ef03

    Change-Id: Ic9ff7d53321c6c7b39f32dc043050f1467a1233d

commit 934213ef03d59624ae8d9e50937ca642156ec90a
Merge: 1e5b0037f 889efd5b9
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Tue Nov 27 08:19:03 2018 -0800

    Fix potential usage of freed memory in btif_hl_proc_sdp_query_cfm
    am: 889efd5b91

    Change-Id: If3ae150367def015874bebb60c2fca763f01133f

commit 2cb7fadb950a6111aa6c9303771d4db8ee8b364f
Merge: 3c0a7052f 9a2ade993
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Mon Nov 26 21:10:41 2018 +0000

    Merge "HFP: support pause character "Comma" in acceptable dial digit string" into pi-dev

commit 889efd5b9165ed7641fcd75eabbbef56be2ef5df
Author: Jakub Pawlowski <jpawlowski@google.com>
Date:   Tue Nov 20 22:31:31 2018 +0100

    Fix potential usage of freed memory in btif_hl_proc_sdp_query_cfm

    Bug: 116222069
    Test: compilation
    Change-Id: Iebe2c500dfc2806ca321fdcd170e20c680619d4d

commit f3681c8616af4d052c410ba3e88747541a974bf5
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Nov 2 20:32:14 2018 +0800

    DO NOT MERGE Separate SDP procedure from bonding state (1/2)

    - Do not stay in bonding state if the device is paried but still
      discovering service.
    - Report BOND_BONDED to Java after authentication is completed.
    - Report empty UUID to Java if a classic Bluetooth device SDP
      failed while pairing.
    - Hold BOND_BONDED intent util SDP is findished.
    - Only accept profile connection for the device is at bonded
      state. Any attempt to connect while bonding would potentially
      lead to an unauthorized connection.

    Bug: 79703832
    Test: runtest bluetooth
    Change-Id: I023713e07308bfc0e5bb8d67f386bcc50f6a0f85
    (cherry picked from commit 122e115b87fe98ca5e5e65b9765c146f9e52b65e)

commit 09894418589479875443037c45e596824da90928
Author: Ugo Yu <ugoyu@google.com>
Date:   Fri Nov 2 19:47:21 2018 +0800

    DO NOT MERGE Separate SDP procedure from bonding state (1/2)

    - Do not stay in bonding state if the device is paried but still
      discovering service.
    - Report BOND_BONDED to Java after authentication is completed.
    - Report empty UUID to Java if a classic Bluetooth device SDP
      failed while pairing.
    - Hold BOND_BONDED intent util SDP is findished.
    - Only accept profile connection for the device is at bonded
      state. Any attempt to connect while bonding would potentially
      lead to an unauthorized connection.

    Bug: 79703832
    Test: runtest bluetooth, regression test
    Change-Id: I023713e07308bfc0e5bb8d67f386bcc50f6a0f85
    Merged-In: I023713e07308bfc0e5bb8d67f386bcc50f6a0f85
    (cherry picked from commit 122e115b87fe98ca5e5e65b9765c146f9e52b65e)

commit bbe8c6125263e40ce8be2f6c6272cded73dafbfd
Merge: 6d1c4974b 4e26a1fa5
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Nov 20 09:11:25 2018 +0000

    [automerger] DO NOT MERGE Separate SDP procedure from bonding state (1/2) am: edd7e731ed am: 279c2a1910 am: c29c3aa408 am: 70ab44a424 skipped: 4e26a1fa5d

    Change-Id: Iaf03c4cdf1ff19989e6d249a1be04ae57346aba0

commit 4e26a1fa5d6e0555de05d566785f52768f4ea5b0
Merge: 9c2fb57ce 70ab44a42
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Nov 20 09:11:21 2018 +0000

    [automerger] DO NOT MERGE Separate SDP procedure from bonding state (1/2) am: edd7e731ed am: 279c2a1910 am: c29c3aa408 am: 70ab44a424

    Change-Id: I0ec2f3cf5ff12ff8953647dc4dbf254fa4573f8b

commit 70ab44a42475fd14c7c9d863f06dfe579bd9e6e9
Merge: c96313fb2 c29c3aa40
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Nov 20 09:11:18 2018 +0000

    [automerger] DO NOT MERGE Separate SDP procedure from bonding state (1/2) am: edd7e731ed am: 279c2a1910 am: c29c3aa408

    Change-Id: I08534e15fd3a1ac53d666a9d27b6f3a30200e065

commit c29c3aa408406d29c6e0e7b1c22a6297ec7f1b1a
Merge: f9606e1d8 279c2a191
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Nov 20 09:11:15 2018 +0000

    [automerger] DO NOT MERGE Separate SDP procedure from bonding state (1/2) am: edd7e731ed am: 279c2a1910

    Change-Id: Ie051800f6ad61b7f7d14dd41f56b19848f38e5fb

commit 279c2a1910c37ae5399bc092b39d6ec98ac6c4bf
Merge: 9805ed7a7 edd7e731e
Author: Android Build Merger (Role) <noreply-android-build-merger@google.com>
Date:   Tue Nov 20 09:11:11 2018 +0000

    [automerger] DO NOT MERGE Separate SDP procedure from bonding state (1/2) am: edd7e731ed

    Change-Id: I1db76ed30b73630aa44839271fbb654ce533c17c

commit edd7e731edad067fe08b0623be6b2745bf81a445
Author: Ugo Yu <ugoyu@google.com>
Date:   Tue Oct 30 15:10:35 2018 +0800

    DO NOT MERGE Separate SDP procedure from bonding state (1/2)

    - Do not stay in bonding state if the device is paried but still
      discovering service.
    - Report BOND_BONDED to Java after authentication is completed.
    - Report empty UUID to Java if a classic Bluetooth device SDP
      failed while pairing.
    - Hold BOND_BONDED intent util SDP is findished.
    - Only accept profile connection for the device is at bonded
      state. Any attempt to connect while bonding would potentially
      lead to an unauthorized connection.

    Bug: 79703832
    Test: runtest bluetooth, regression test.
    Change-Id: I023713e07308bfc0e5bb8d67f386bcc50f6a0f85
    (cherry picked from commit 122e115b87fe98ca5e5e65b9765c146f9e52b65e)

commit 730fcaa764b5060e570fd6cadfc7a9180cbca517
Merge: b0ea5e427 163dec2ae
Author: Hansong Zhang <hsz@google.com>
Date:   Mon Nov 5 18:03:36 2018 +0000

    Merge "DO NOT MERGE HFP: Check AT command buffer boundary during parsing" into nyc-dev

commit b0ea5e427bfca001b00dd2d219f4740bd2db2db7
Merge: fefbf9605 a4a11e198
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Mon Nov 5 17:16:34 2018 +0000

    Merge "DO NOT MERGE: HH: Check parameter length in bta_hh_ctrl_dat_act" into nyc-dev

commit fefbf9605432e68b16c560ad185c516fed939d72
Merge: 5b3dc18e3 2091fe750
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Fri Nov 2 22:47:59 2018 +0000

    Merge "DO NOT MERGE: SDP: Check p_end in save_attr_seq and add_attr" into nyc-dev

commit 5b3dc18e362f3ca246969cb7cba701555d8d2ace
Merge: ad4098c34 840f70ca1
Author: TreeHugger Robot <treehugger-gerrit@google.com>
Date:   Fri Nov 2 22:43:43 2018 +0000

    Merge "DO NOT MERGE: MCAP: Check response length in mca_ccb_hdl_rsp" into nyc-dev

commit 9a2ade993445ba334df688285deb4dbf024d082b
Author: jonerlin <jonerlin@google.com>
Date:   Tue Oct 30 11:32:21 2018 +0800

    HFP: support pause character "Comma" in acceptable dial digit string

    * To avoid bt stack filter MO call initiated from HF end with by sending dial command with modifier character "," (Comma)

    Bug: 118401445
    Test: making MO call from carkit/lab-equipment by sending ATD command with phone number plus comma plus DTMF digitals
    Change-Id: I931147ee970ff30f61913d0de138b53eba40eaf5

Change-Id: I18db90667f9477244c3df2a4bbc0efb5e78b5182
---
 bta/sys/utl.cc              |  3 ++-
 btif/avrcp/avrcp_service.cc | 19 ++++++++++---------
 btif/src/bluetooth.cc       |  2 --
 profile/avrcp/device.cc     |  2 ++
 profile/avrcp/device.h      |  6 ++++++
 stack/avrc/avrc_api.cc      |  6 +++++-
 6 files changed, 25 insertions(+), 13 deletions(-)

diff --git a/bta/sys/utl.cc b/bta/sys/utl.cc
index 92318d43c..544f91e8c 100644
--- a/bta/sys/utl.cc
+++ b/bta/sys/utl.cc
@@ -242,7 +242,8 @@ bool utl_isintstr(const char* p_s) {
  ******************************************************************************/
 bool utl_isdialchar(const char d) {
   return (((d >= '0') && (d <= '9')) || (d == '*') || (d == '+') ||
-          (d == '#') || (d == ';') || ((d >= 'A') && (d <= 'C')) ||
+          (d == '#') || (d == ';') || (d == ',') ||
+          ((d >= 'A') && (d <= 'C')) ||
           ((d == 'p') || (d == 'P') || (d == 'w') || (d == 'W')));
 }
 
diff --git a/btif/avrcp/avrcp_service.cc b/btif/avrcp/avrcp_service.cc
index 8c22365b2..5a058f4af 100644
--- a/btif/avrcp/avrcp_service.cc
+++ b/btif/avrcp/avrcp_service.cc
@@ -346,10 +346,11 @@ void AvrcpService::SendMediaUpdate(bool track_changed, bool play_state,
 
   // This function may be called on any thread, we need to make sure that the
   // device update happens on the main thread.
-  for (auto device : instance_->connection_handler_->GetListOfDevices()) {
-    do_in_bta_thread(FROM_HERE, base::Bind(&Device::SendMediaUpdate,
-                                           base::Unretained(device.get()),
-                                           track_changed, play_state, queue));
+  for (const auto& device :
+       instance_->connection_handler_->GetListOfDevices()) {
+    do_in_bta_thread(FROM_HERE,
+                     base::Bind(&Device::SendMediaUpdate, device.get()->Get(),
+                                track_changed, play_state, queue));
   }
 }
 
@@ -361,11 +362,11 @@ void AvrcpService::SendFolderUpdate(bool available_players,
             << " uids=" << uids;
 
   // Ensure that the update is posted to the correct thread
-  for (auto device : instance_->connection_handler_->GetListOfDevices()) {
-    do_in_bta_thread(
-        FROM_HERE,
-        base::Bind(&Device::SendFolderUpdate, base::Unretained(device.get()),
-                   available_players, addressed_players, uids));
+  for (const auto& device :
+       instance_->connection_handler_->GetListOfDevices()) {
+    do_in_bta_thread(FROM_HERE,
+                     base::Bind(&Device::SendFolderUpdate, device.get()->Get(),
+                                available_players, addressed_players, uids));
   }
 }
 
diff --git a/btif/src/bluetooth.cc b/btif/src/bluetooth.cc
index dc59046c8..6e86ce780 100644
--- a/btif/src/bluetooth.cc
+++ b/btif/src/bluetooth.cc
@@ -326,8 +326,6 @@ static void dump(int fd, const char** arguments) {
 #if (BTSNOOP_MEM == TRUE)
   btif_debug_btsnoop_dump(fd);
 #endif
-
-  close(fd);
 }
 
 static void dumpMetrics(std::string* output) {
diff --git a/profile/avrcp/device.cc b/profile/avrcp/device.cc
index 9b25fac03..7f8ecae3a 100644
--- a/profile/avrcp/device.cc
+++ b/profile/avrcp/device.cc
@@ -52,6 +52,8 @@ void Device::RegisterInterfaces(MediaInterface* media_interface,
   volume_interface_ = volume_interface;
 }
 
+base::WeakPtr<Device> Device::Get() { return weak_ptr_factory_.GetWeakPtr(); }
+
 bool Device::IsActive() const {
   return address_ == a2dp_interface_->active_peer();
 }
diff --git a/profile/avrcp/device.h b/profile/avrcp/device.h
index 5ef886d26..b1f34208c 100644
--- a/profile/avrcp/device.h
+++ b/profile/avrcp/device.h
@@ -55,6 +55,12 @@ class Device {
       uint16_t ctrl_mtu, uint16_t browse_mtu);
   virtual ~Device() = default;
 
+  /**
+   * Gets a weak pointer to this device that is invalidated when the device is
+   * disconnected.
+   */
+  base::WeakPtr<Device> Get();
+
   const RawAddress& GetAddress() const { return address_; };
 
   /**
diff --git a/stack/avrc/avrc_api.cc b/stack/avrc/avrc_api.cc
index 53c25a946..e2f109a31 100644
--- a/stack/avrc/avrc_api.cc
+++ b/stack/avrc/avrc_api.cc
@@ -1150,7 +1150,11 @@ uint16_t AVRC_MsgReq(uint8_t handle, uint8_t label, uint8_t ctype,
     AVRC_CO_ID_TO_BE_STREAM(p_data, AVRC_CO_METADATA);
   } else {
     chk_frag = false;
-    peer_mtu = AVCT_GetBrowseMtu(handle);
+    if (p_pkt->layer_specific == AVCT_DATA_BROWSE) {
+      peer_mtu = AVCT_GetBrowseMtu(handle);
+    } else {
+      peer_mtu = AVCT_GetPeerMtu(handle);
+    }
     if (p_pkt->len > (peer_mtu - AVCT_HDR_LEN_SINGLE)) {
       AVRC_TRACE_ERROR(
           "%s bigger than peer mtu (p_pkt->len(%d) > peer_mtu(%d-%d))",
-- 
2.17.1

